clang/test/SemaSYCL/intel-fpga-local.cpp

```diff
@@ -48,8 +48,6 @@ void check_ast()
   //CHECK: VarDecl{{.*}}private_copies
   //CHECK: IntelFPGAMemoryAttr{{.*}}Implicit
   //CHECK: IntelFPGAPrivateCopiesAttr
-  //CHECK-NEXT: ConstantExpr
```

Comment:
This looks like a regression -- we're no longer retaining the attribute argument in the AST (so you can't AST match this value either). I think the reason why it's no longer showing up is because the given expression is no longer being converted into a constant expression -- we evaluate it as such, but then pass in the original expression (not the converted constant expression) when creating the attribute.

Reply:
Good question! I'm thinking out loud here...

I could be wrong, but I'm pretty sure that `addIntelSYCLSingleArgFunctionAttr()` is calling `getIntegerConstantExpr()`, which goes through all the trouble of converting the expression into a constant expression but then throws the expr results away. So if we're going to go to the trouble of converting the expression to a constant expression, it seems reasonable to store the converted results if they're useful. AST dumping is an example of a case where it's somewhat useful (seeing the constant expression value is useful for humans, but more interestingly, there's a JSON dump that's used by other tools for automation and this would prevent those tools from having to do the work of evaluating the expression to get its result). AST matching also gets somewhat easier in that you can write an AST matcher that checks whether the argument has a specific value (but this is somewhat less compelling because AST matching on attribute arguments is really brand new functionality that's got a lot of rough edges). I think the most compelling reason is for ease of implementation within Clang itself -- if we convert the arguments to be a constant expression and store that in the semantic attribute, then users of the semantic attribute can rely on the expression having already been evaluated and more directly get the results without having to evaluate the constant expression a second time (such as when doing codegen).

So I think we should consistently store a `ConstantExpr` as the attribute argument, and then audit the places where we try to get the value out of the semantic attribute to see if there are places where we can remove duplicate effort (by calling `ConstantExpr::Int64Result()` or `ConstantExpr::APValueResult()` instead of re-evaluating the expression). WDYT?

Reply:
I think `AddOneConstantValueAttr` does the constant generation, not `addIntelSYCLSingleArgFunctionAttr `. I believe the difference lies in `VerifyIntegerConstantExpression `. This is called in `AddOneConstantValueAttr`. We're seeing it removed in this patch because I changed calls to `AddOneConstantValueAttr` with `addIntelSYCLSingleArgFunctionAttr ` because the formed has a call to `checkRangedIntegralArgument` which uses the additional members I'm removing in Attribute definition

Reply:
> So I think we should consistently store a `ConstantExpr` as the attribute argument, and then audit the places where we try to get the value out of the semantic attribute to see if there are places where we can remove duplicate effort (by calling `ConstantExpr::Int64Result()` or `ConstantExpr::APValueResult()` instead of re-evaluating the expression). WDYT?

Sounds good to me. I'll make the change. Thanks for the feedback!

Reply:
I verified that my understanding was correct. `addIntelSYCLSingleArgFunctionAttr()` calls `E->getIntegerConstantExpr()` on the given expression, and `Expr::getIntegerConstantExpr()` evaluates the constant expression but throws away the expression results: https://github.com/intel/llvm/blob/sycl/clang/lib/AST/ExprConstant.cpp#L15518

Meanwhile, `AddOneConstantValueAttr()` calls `checkRangedIntegralArgument()` which calls `VerifyIntegerConstantExpression()` to convert the expression to a constant expression. This constant expression is returned to the caller, which resets the given expression to be the constant expression.

So either way, we're converting the expression to a constant expression, but only in one case are we throwing that useful work away (and I claim we shouldn't throw that work away but instead store it in the semantic attribute and make use of it elsewhere).

Reply:
Ok Thank you!